### PR TITLE
Adds simple transaction commands

### PR DESCRIPTION
**Pull request by Till-Later**

### CODE REVIEW

**Comment by mrks** [comment]
>  it is actually always valid to commit or Rollback a transaction even though you never found a BEGIN.

That is way out of the parser's responsibility. A client might send multiple statements, so the parser does not necessarily see the whole picture. Also, a transaction might fail in-flight, which the parser cannot know about.

File: src/parser/bison_parser.y [hunk 1]
```diff
@@ -315,6 +318,7 @@ preparable_statement:
 	|	update_statement { $$ = $1; }
 	|	drop_statement { $$ = $1; }
 	|	execute_statement { $$ = $1; }
+	|   transaction_statement { $$ = $1; }
```
Comment (mrks) [inline src/parser/bison_parser.y]: whitespace

Comment (mrks) [inline src/parser/bison_parser.y]: Are the transaction commands part of a preparable statement? Have a look at https://www.contrib.andrew.cmu.edu/~shadow/sql/sql1992.txt - especially <SQL transaction statement> and <SQL dynamic statement> in 12.5

File: src/sql/statements.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ namespace hsql {
     delete expr;
   }
 
-  // DropStatament
+  // DropStatement
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/sql/statements.cpp]: fixed a typo

File: src/sql/statements.cpp [hunk 3]
```diff
@@ -117,6 +117,14 @@ namespace hsql {
     free(name);
   }
 
+  // TransactionStatement
+  TransactionStatement::TransactionStatement(TransactionAction action)
+    : SQLStatement(kStmtTransaction)
+    , action(action) {};
```
Comment (mrks) [inline src/sql/statements.cpp]: format

**Comment by Dencrash** [comment]
> > it is actually always valid to commit or Rollback a transaction even though you never found a BEGIN.
> 
> That is way out of the parser's responsibility. A client might send multiple statements, so the parser does not necessarily see the whole picture. Also, a transaction might fail in-flight, which the parser cannot know about.

Makes sense.

File: src/parser/bison_parser.y (hunk 1 above)
Comment (klauck) [reply src/parser/bison_parser.y]: Yes (in the SQL standard), see 17.6 on page 441.
Note, our current grammar cannot be mapped directly to the standard.

**Comment by youri-k** [comment]
We finished implementing the PR feedback. @Dencrash @mrks could review the changes once again. Thanks :slightly_smiling_face:

File: src/util/sqlhelper.cpp [hunk 4]
```diff
@@ -261,6 +261,18 @@ namespace hsql {
     }
   }
 
+  void printTransactionStatementInfo(const TransactionStatement* stmt, uintmax_t numIndent) {
+    inprint("TransactionStatement", numIndent);
+    switch (stmt->command){
+    case kBeginTransaction: inprint("BEGIN", numIndent + 1);
+      break;
```
Comment (mrks) [inline src/util/sqlhelper.cpp]: format: inpritn should be in its own line

File: src/parser/bison_parser.y [hunk 5]
```diff
@@ -180,13 +181,15 @@ int yyerror(YYLTYPE* llocp, SQLParserResult* result, yyscan_t scanner, const cha
...
 %token TRUE FALSE
+%token TRANSACTION BEGIN COMMIT ROLLBACK
 
 /*********************************
  ** Non-Terminal types (http://www.gnu.org/software/bison/manual/html_node/Type-Decl.html)
  *********************************/
 %type <stmt_vec>	    statement_list
 %type <statement> 	    statement preparable_statement
 %type <exec_stmt>	    execute_statement
+%type <transaction_stmt> transaction_statement
```
Comment (mrks) [inline src/parser/bison_parser.y]: whitespace

File: src/sql/statements.cpp [hunk 6]
```diff
@@ -117,6 +117,14 @@ namespace hsql {
     free(name);
   }
 
+  // TransactionStatement
+  TransactionStatement::TransactionStatement(TransactionCommand command)
+    : SQLStatement(kStmtTransaction)
+    , command(command) {};
```
Comment (mrks) [inline src/sql/statements.cpp]: comma in previous line

**Comment by youri-k** [comment]
Just updated the formatting. Thanks for the hints :slightly_smiling_face: 

Unfortunately, in the `bison_parser.y` the indentation is done with both tab and spaces. Therefore, the appearance is dependent on the tab size setting.

**Pull request merged**